Fix _empty_per_channel_affine_quantized to be less hacky

dzhulgakov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26676 Allow per-channel QTensor accept any floating type for scales
* #26675 Per-channel quantized tensor to have only a single axis
* **#26243 Fix _empty_per_channel_affine_quantized to be less hacky**

This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

jamesr66a commented on aten/src/ATen/native/README.md:
```diff
@@ -99,7 +99,9 @@ signature.
 differently than another example, check first and see if one is a
-factory while another is not.
+factory while another is not. In some rare cases, factory function might have a
+tensor argument. In this case it'd be marked with 'category_override: factory'
```
`it'd be marked` -> `mark it`

jamesr66a commented:
Can you post a diff of the codegen'ed files before and after this change?

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -142,10 +142,10 @@ Tensor per_channel_affine_qtensor_cpu(
     const Tensor& zero_points,
     IntArrayRef axis) {
-  Tensor dst = at::_empty_per_channel_affine_quantized_like(
+  Tensor dst = at::_empty_per_channel_affine_quantized(
```
what should be the order of arguments?
should we put size in the end so it is closer to options?

dzhulgakov replied:
I think sizes should be first - similarly to how we do it for at::full or at::rand

raghuramank100 reviewed (commented):
Looks good to me, will let folks more familiar with this part of the code approve.

gchanan reviewed (commented):
ya I'd like to see the codegen changes.

I think the original issue is we didn't know how to dispatch these things.  Because we usually dispatch based on the tensor argument but in this case, I'm not even sure what we were dispatching on.  I also think @ezyang's recent multi dispatch changes might have affected this, because I know he was changing how "new" functions worked.

dzhulgakov commented:
Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

ezyang commented:
The conflict with my changes looks very mild and I can help resolve the changes if I land before this one. (Or if you land before me, I can fix it in my patch, but I'm landing right now and hope it will work this time ;)

Commit:
Update on "Fix _empty_per_channel_affine_quantized to be less hacky"


This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

[ghstack-poisoned]

Commit:
Update on "Fix _empty_per_channel_affine_quantized to be less hacky"


This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -991,7 +991,9 @@
 
-- func: _empty_per_channel_affine_quantized_like(Tensor self, Tensor zero_points, int[] size, int[] axis, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=contiguous_format) -> Tensor
+# it's a factory function receiving a tensor argument, thus overriding explicitly
+- func: _empty_per_channel_affine_quantized(int[] size, *, Tensor scales, Tensor zero_points, int[] axis, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None, MemoryFormat? memory_format=contiguous_format) -> Tensor
```
I think this is all a pre-existing issue with TensorOptions, but I don't see how the dispatch makes sense here.

For example, if I don't specify a dtype, I won't try to dispatch to QuantizedCPU and presumably I'll just get some crappy error message about _empty_per_channel_affine_quantized not being implemented for CPU backend.

But really...dtype is not optional.  So maybe we don't want to do anything until @izdeby breaks apart TensorOptions, but it seems reasonable to provide a CPU overload and to give a real error message here.

dzhulgakov replied:
Good idea, let me fix the error message.Btw, it applies to other qtensor factory functions too

Commit:
Update on "Fix _empty_per_channel_affine_quantized to be less hacky"


This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

[ghstack-poisoned]

Commit:
Update on "Fix _empty_per_channel_affine_quantized to be less hacky"


This is an attempt to fix _empty_per_channel_affine_quantized to be more sane. It's a factory function that nevertheless receives a Tensor argument and it throws the codegen off course.

Before people did a hacky workaround of appending _like to the function name to trick codegen, it also required non-natural argument order.

This PR explicitly allows to override the 'category' of the function to make codegen do the right thing. Now name and the argument order (in C++) make more sense.

Gist with changes to codegen: https://gist.github.com/dzhulgakov/87050fcf7527730bea6483600f8c4897 (from ./tools/git_add_generated_dirs.sh)

Differential Revision: [D17443221](https://our.internmc.facebook.com/intern/diff/D17443221)

[ghstack-poisoned]
